drivers/usb/udc/udc_stm32.c

```diff
@@ -260,6 +243,28 @@ void HAL_PCD_DataOutStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
 {
 	uint32_t rx_count = HAL_PCD_EP_GetRxCount(hpcd, epnum);
 	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct udc_stm32_msg msg = {
```

Comment:
I caught an issue where this check would trigger `ZLP expected` (only appeared on usbotg_hs), as multiple messages arriving before being handled would be merged. Storing the rx length in the msgq fixed this. It doubles the size of the messages, but I dont see a way around it. The cdc_acm test still fails on usbotg_hs.
https://github.com/zephyrproject-rtos/zephyr/blob/c0be51df2fb83f2e772d74b73e10350662af12b8/drivers/usb/udc/udc_common.c#L1058-L1064

Reply:
Nice! I think this could potentially fix a couple of issues seen by @fpistm (particularly on Windows). He'll only be able to answer on monday though.

> The cdc_acm test still fails on usbotg_hs.

Which board are you using for this test ?

Reply:
> Which board are you using for this test ?
Mainly testing using stm32f769i_disco, but i see the same on a portenta_h7. I may be able to test the fs interface on the h7 to verify that it only happens on hs, but the portenta is mounted in a custom board where its a bit more difficult to perform isolated tests.
Just to clarify: when running the cdc_acm sample, descriptors are exchanged without issue, but no characters are echoed back. I am currently doing some debugging with wireshark, i will report a separate issue once i have more context.

Reply:
@topisani 
I experienced the same problem of characters not being echoed back with the stm32h747i_disco board, but it ran smoothly on the nucleo_h743zi board, which means it only occurs on HS.

Reply:
i have now also confirmed that it works on the usbotg_fs interface on stm32f769i_disco. This issue with HS also exists on main.